rust/src/operations/delete.rs

```diff
@@ -0,0 +1,812 @@
+                expr,
+                &current_metadata.partition_columns,
+                &mut expr_properties,
+            )?;
+            excute_non_empty_expr(
+                snapshot,
+                object_store.clone(),
+                &state,
+                expr,
+                &mut metrics,
+                writer_properties,
+                expr_properties,
+            )
+            .await?
+        }
+        None => (Vec::<Add>::new(), snapshot.files().to_owned()),
+    };
+
+    let predicate = predicate.unwrap_or(Expr::Literal(ScalarValue::Boolean(Some(true))));
+
+    let deletion_timestamp = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_millis() as i64;
+
+    let mut actions: Vec<Action> = add_actions.into_iter().map(Action::add).collect();
+    metrics.num_removed_files = to_delete.len();
+    metrics.num_added_files = actions.len();
+
+    for action in to_delete {
+        actions.push(Action::remove(Remove {
+            path: action.path,
+            deletion_timestamp: Some(deletion_timestamp),
+            data_change: true,
+            extended_file_metadata: Some(true),
+            partition_values: Some(action.partition_values),
+            size: Some(action.size),
+            tags: None,
+        }))
+    }
+
+    metrics.execution_time_ms = Instant::now().duration_since(exec_start).as_micros();
+
+    // Do not make a commit when there are zero updates to the state
+    if !actions.is_empty() {
+        let operation = DeltaOperation::Delete {
+            predicate: Some(predicate.canonical_name()),
+        };
+        commit(
+            object_store.as_ref(),
+            &actions,
+            operation,
+            snapshot,
+            app_metadata,
+        )
+        .await?;
+    }
+
+    Ok(metrics)
+}
+
+impl std::future::IntoFuture for DeleteBuilder {
+    type Output = DeltaResult<(DeltaTable, DeleteMetrics)>;
+    type IntoFuture = BoxFuture<'static, Self::Output>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let this = self;
+
+        Box::pin(async move {
+            let writer_properties = this
+                .writer_properties
+                .unwrap_or_else(|| WriterProperties::builder().build());
+
+            let state = this.state.unwrap_or_else(|| {
+                let session = SessionContext::new();
+
+                // If a user provides their own their DF state then they must register the store themselves
+                register_store(this.store.clone(), session.runtime_env());
+
+                session.state()
+            });
+
+            let metrics = execute(
+                this.predicate,
+                this.store.clone(),
+                &this.snapshot,
+                state,
+                writer_properties,
+                None,
+            )
+            .await?;
+            let mut table = DeltaTable::new_with_state(this.store, this.snapshot);
+            table.update().await?;
```

Comment:
Here we can probably avoid some io by returning the actions and version from `execute` and updating the new state directly. Essentially something like.

```rs
this.snapshot.merge(DeltaTableState::from_actions(actions, version)?, true, true);
```

Reply:
I implemented this and I imagine we would want to use this pattern for other operations. Should we return the version and actions in an optional? Currently I return an empty list + the current version if there are no changes.